### PR TITLE
nvgpu concat split

### PR DESCRIPTION
**Pull request by fengleitian**
https://github.com/NervanaSystems/ngraph/issues/1867

split large concat to small part to avoid parameter space overflow. 

fail example:
NGRAPH_TF_BACKEND=GPU python cifar10_main.py --resnet_size=20 --model_dir --batch_size=512

### CODE REVIEW

File: test/backend_test.in.cpp [hunk 1]
```diff
@@ -440,6 +440,41 @@ NGRAPH_TEST(${BACKEND_NAME}, concat_matrix_int64)
               read_vector<int64_t>(result));
 }
 
+//Add this test to cover paramter space overflow:
```
Comment (rkimballn1) [inline test/backend_test.in.cpp]: Can you put a space after `//` to match other comments

Comment (fengleitian) [reply test/backend_test.in.cpp]: fixed.

File: src/ngraph/runtime/gpu/cuda_emitter.hpp [hunk 2]
```diff
@@ -174,7 +174,7 @@ namespace ngraph
                                          NVShape input_dilation,
                                          NVDiff input_pad_below);
 
-                size_t build_concat(const std::vector<std::string>& dtypes,
+                size_t build_concat(const std::string& dtypes,
```
Comment (gcwenger) [inline src/ngraph/runtime/gpu/cuda_emitter.hpp]: dtypes->dtype for consistency w/ implementation

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 3]
```diff
@@ -104,64 +104,109 @@ size_t runtime::gpu::CUDAEmitter::build_concat(const std::vector<std::string>& d
...
-    if (compiled_kernel == nullptr)
+    // recompile the kernel and then create the primutive
+    size_t split_input_size = 128;
+    size_t residue = input_num % split_input_size;
+    std::stringstream kernel_name_1;
+    std::stringstream kernel_name_2;
+    kernel_name_1 << "concat_" << dtype << "_r_" << split_input_size;
+    kernel_name_2 << "concat_" << dtype << "_r_" << residue;
+    auto compiled_kernel_1 = m_ctx->compiled_kernel_pool->get(kernel_name_1.str());
+    if (compiled_kernel_1 == nullptr && input_num >= split_input_size)
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: @gcwenger pointed this out to me as I missed this check, ignore my above comment about double compilation.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp [hunk 4]
```diff
@@ -51,7 +51,7 @@ namespace ngraph
 
                 static void get_concat_op(codegen::CodeWriter& writer,
                                           const std::string& name,
-                                          const std::vector<std::string>& data_types,
+                                          const std::string& data_types,
```
Comment (gcwenger) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp]: data_types->data_type for consistency w/ implementation.

File: test/backend_test.in.cpp [hunk 5]
```diff
@@ -441,6 +441,41 @@ NGRAPH_TEST(${BACKEND_NAME}, concat_matrix_int64)
...
 
+// Add this test to cover paramter space overflow:
+// cuda kernel parameter space have limit, if there is large number of parameters,
+// there will be overflow for parameter space.
+NGRAPH_TEST(${BACKEND_NAME}, concat_vector_large)
+{
+    Shape shape_a{1};
+    NodeVector inputs;
+    op::ParameterVector inputs_param;
+    for (uint32_t i = 0; i < 999; i++)
```
Comment (gcwenger) [inline test/backend_test.in.cpp]: This only tests case when number of inputs is > your split point. Would be good to test exactly same number as your split point (concat_vector test already tests < your split point). Since this is reference case driven, you could do this as a parameterized test (I added support for this and can describe mechanism) and generate 2 or more variants.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 6]
```diff
@@ -104,64 +104,109 @@ size_t runtime::gpu::CUDAEmitter::build_concat(const std::vector<std::string>& d
     // check if the kernel has already been compiled. if so, create
     // a launch primitive for it based on the input tensor shape
     // but do not recompile the kernel. otherwise, do it all:
-    // recompile the kernel and then create the primitive
-    auto compiled_kernel = m_ctx->compiled_kernel_pool->get(kernel_name.str());
-    if (compiled_kernel == nullptr)
+    // recompile the kernel and then create the primutive
+    size_t split_input_size = 128;
```
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: max 2^x number we can do without overflow.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 7]
```diff
@@ -104,64 +104,109 @@ size_t runtime::gpu::CUDAEmitter::build_concat(const std::vector<std::string>& d
...
+            auto compiled_kernel =
+                (args_list.size() == split_input_size + 7) ? compiled_kernel_1 : compiled_kernel_2;
+            CUDA_SAFE_CALL(cuLaunchKernel(*compiled_kernel.get(),
+                                          split_aligned_grid_size_x[n],
+                                          1,
+                                          1, // grid dim
+                                          block_size_x,
+                                          1,
+                                          1, // block dim
+                                          0,
```
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: in another PR, we don't have a stream support yet.

File: src/ngraph/runtime/gpu/cuda_emitter.hpp (hunk 2 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.hpp]: changed.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp (hunk 4 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.hpp]: changed.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: I thought I read sometime back that the parameter limit was 4KB of constant mem.

[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#function-parameters

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: Maybe it's not so hard to add here but we can do it generally later and avoid the RT cost with some RAII support for streams as with the cudnn descriptors.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: Ye, 4096 Bytes, but each parameters is about 8 bytes or more.
